# Data Wrangling Project - 1.

In [13]:
# import libraries
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer
import pandas as pd
import requests
import copy

## Gather

In [2]:
# Downlaod image predictions
r = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv')
with open('image-predictions.tsv','wb') as f:
    f.write(r.content)

In [20]:
#### twitter-archive-enhanced.csv file
df_1 = pd.read_csv('twitter-archive-enhanced.csv')
df_1.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
41    884441805382717440                    NaN                  NaN   
1085  738184450748633089                    NaN                  NaN   
553   804026241225523202                    NaN                  NaN   
600   798673117451325440                    NaN                  NaN   
1890  674767892831932416                    NaN                  NaN   

                      timestamp  \
41    2017-07-10 15:58:53 +0000   
1085  2016-06-02 01:44:22 +0000   
553   2016-11-30 18:16:08 +0000   
600   2016-11-15 23:44:44 +0000   
1890  2015-12-10 01:49:36 +0000   

                                                 source  \
41    <a href="http://twitter.com/download/iphone" r...   
1085  <a href="http://twitter.com/download/iphone" r...   
553   <a href="http://twitter.com/download/iphone" r...   
600   <a href="http://twitter.com/download/iphone" r...   
1890  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
41    I present to you, Pup in Hat. Pup in Hat is gr...                  NaN   
1085  For the last time, we only rate dogs. Pls stop...                  NaN   
553   This is Bo. He's going to make me cry. 13/10 p...                  NaN   
600   RT @dog_rates: I shall call him squishy and he...         6.755011e+17   
1890  This pup was carefully tossed to make it look ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
41                         NaN                        NaN   
1085                       NaN                        NaN   
553                        NaN                        NaN   
600               4.196984e+09  2015-12-12 02:23:01 +0000   
1890                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
41    https://twitter.com/dog_rates/status/884441805...                14   
1085  https://twitter.com/dog_rates/status/738184450...                 9   
553   https://twitter.com/dog_rates/status/804026241...                13   
600   https://twitter.com/dog_rates/status/675501075...                13   
1890  https://twitter.com/dog_rates/status/674767892...                12   

      rating_denominator  name doggo floofer pupper puppo  
41                    10  None  None    None   None  None  
1085                  10  None  None    None   None  None  
553                   10    Bo  None    None   None  None  
600                   10  None  None    None   None  None  
1890                  10  None  None    None   None  None

In [ ]:
# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = df_1.tweet_id.values
len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'wb') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)


In [4]:
# Flat file reading using pandas Dataframe
df_2 = pd.read_csv('image-predictions.tsv', sep= '\t')
df_2.head(30)

tweet_id                                          jpg_url  \
0   666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1   666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2   666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3   666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4   666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5   666050758794694657  https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6   666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7   666055525042405380  https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8   666057090499244032  https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9   666058600524156928  https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10  666063827256086533  https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11  666071193221509120  https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12  666073100786774016  https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13  666082916733198337  https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14  666094000022159362  https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15  666099513787052032  https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16  666102155909144576  https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17  666104133288665088  https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18  666268910803644416  https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19  666273097616637952  https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20  666287406224695296  https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21  666293911632134144  https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22  666337882303524864  https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23  666345417576210432  https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24  666353288456101888  https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25  666362758909284353  https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26  666373753744588802  https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27  666396247373291520  https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28  666407126856765440  https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29  666411507551481857  https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   

    img_num                           p1   p1_conf  p1_dog  \
0         1       Welsh_springer_spaniel  0.465074    True   
1         1                      redbone  0.506826    True   
2         1              German_shepherd  0.596461    True   
3         1          Rhodesian_ridgeback  0.408143    True   
4         1           miniature_pinscher  0.560311    True   
5         1         Bernese_mountain_dog  0.651137    True   
6         1                   box_turtle  0.933012   False   
7         1                         chow  0.692517    True   
8         1                shopping_cart  0.962465   False   
9         1             miniature_poodle  0.201493    True   
10        1             golden_retriever  0.775930    True   
11        1                Gordon_setter  0.503672    True   
12        1                 Walker_hound  0.260857    True   
13        1                          pug  0.489814    True   
14        1                   bloodhound  0.195217    True   
15        1                        Lhasa  0.582330    True   
16        1               English_setter  0.298617    True   
17        1                          hen  0.965932   False   
18        1             desktop_computer  0.086502   False   
19        1            Italian_greyhound  0.176053    True   
20        1                  Maltese_dog  0.857531    True   
21        1             three-toed_sloth  0.914671   False   
22        1                           ox  0.416669   False   
23        1             golden_retriever  0.858744    True   
24        1                     malamute  0.336874    True   
25        1                   guinea_pig  0.996496   False   
26        1  s

#### Extract tweet id, retweets and tweet favorites from tweet-json.txt file to a pandas Dataframe

In [11]:
tweet_list = []

with open('tweet-json.txt', encoding = 'utf-8') as f:
    each_tweet = f.readlines()
    for i in range(len(each_tweet)):
        tweet = json.loads(each_tweet[i])
        id = tweet['id']
        retweet = tweet['retweet_count']
        like = tweet['favorite_count']
        tweet_list.append({'tweet_id': id,
                          'retweets':retweet,
                          'tweet_favorites': like,
                          })
df_3 = pd.DataFrame(tweet_list, columns = ['tweet_id', 'retweets', 'tweet_favorites'])        

In [12]:
# check the final output
df_3.head()

tweet_id  retweets  tweet_favorites
0  892420643555336193      8853            39467
1  892177421306343426      6514            33819
2  891815181378084864      4328            25461
3  891689557279858688      8964            42908
4  891327558926688256      9774            41048

## Assess

In [22]:
# copy twitter-archive-enhanced file
archive = df_1.copy()   

In [25]:
archive.sample(100)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1107  734776360183431168                    NaN                  NaN   
651   792773781206999040                    NaN                  NaN   
1216  714982300363173890                    NaN                  NaN   
2078  670832455012716544                    NaN                  NaN   
501   813096984823349248                    NaN                  NaN   
1332  705475953783398401                    NaN                  NaN   
1168  721001180231503872                    NaN                  NaN   
1979  672980819271634944                    NaN                  NaN   
64    879674319642796034           8.795538e+17         3.105441e+09   
1532  690005060500217858                    NaN                  NaN   
1542  689599056876867584                    NaN                  NaN   
885   760153949710192640                    NaN                  NaN   
873   761334018830917632                    NaN                  NaN   
688   787810552592695296                    NaN                  NaN   
704   785872687017132033                    NaN                  NaN   
1855  675531475945709568                    NaN                  NaN   
2164  669371483794317312                    NaN                  NaN   
1536  689905486972461056                    NaN                  NaN   
620   796125600683540480                    NaN                  NaN   
154   862096992088072192                    NaN                  NaN   
2147  669753178989142016                    NaN                  NaN   
256   843981021012017153                    NaN                  NaN   
401   824663926340194305                    NaN                  NaN   
1746  679132435750195208                    NaN                  NaN   
1365  702684942141153280                    NaN                  NaN   
710   784826020293709826                    NaN                  NaN   
2070  671122204919246848                    NaN                  NaN   
1594  686377065986265092                    NaN                  NaN   
2017  672160042234327040                    NaN                  NaN   
1831  676215927814406144                    NaN                  NaN   
...                  ...                    ...                  ...   
227   848324959059550208                    NaN                  NaN   
282   839239871831150596                    NaN                  NaN   
1815  676613908052996102                    NaN                  NaN   
1614  685307451701334016                    NaN                  NaN   
1916  674307341513269249                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
625   795400264262053889                    NaN                  NaN   
1182  718971898235854848                    NaN                  NaN   
1679  682032003584274432                    NaN                  NaN   
1199  716791146589110272                    NaN                  NaN   
1393  700029284593901568                    NaN                  NaN   
1178  719551379208073216                    NaN                  NaN   
187   856282028240666624                    NaN                  NaN   
870   761672994376806400                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
635   793601777308463104                    NaN                  NaN   
449   819004803107983360                    NaN                  NaN   
1527  690360449368465409                    NaN                  NaN   
290   838150277551247360           8.381455e+17         2.195506e+07   
457   818145370475810820                    NaN                  NaN   
2134  670069087419133954                    NaN                  NaN   
1113  733109485275860992                    NaN                  NaN   
96    873580283840344065                    NaN                  NaN   
827   769212283578875904                    NaN         

In [23]:
archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [26]:
# copy image-predictions
predictions = df_2.copy()

In [28]:
predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [51]:
archive.tweet_id.duplicated().value_counts()


False    2356
Name: tweet_id, dtype: int64

In [46]:
animals = pd.Series(['lama', 'cow', 'lama', 'beetle', 'lama'])
animals.duplicated().value_counts()

False    3
True     2
dtype: int64

In [29]:
predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [53]:
# copy tweeter information file
tweet_info = df_3.copy()
tweet_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
tweet_id           2354 non-null int64
retweets           2354 non-null int64
tweet_favorites    2354 non-null int64
dtypes: int64(3)
memory usage: 55.2 KB


#### Quality

***archive table***
- timestamp is a string format not a datetime
- retweeted_status_id is a float 
- retweeted_status_user_id is a float
- timestamp has validity issue
- retweeted timestamp is a string format not a datetime

***predictions table***
- p1, p2 and p3 algorithm namings are not sometime lowercase
- p1, p2 and p3 algorithm names are separated with underscore
- mission records 2075 not 2356
- prediction output column are needed to be rearranged


***tweet info table***
- missing records 2354 not 2356



#### Tidiness

***archive table***
- timestamp date and time infor in one column


## Clean

In [ ]:
tweet_list = []

with open('tweet-json.txt', encoding = 'utf-8') as f:
    
    each_tweet = f.readlines()
    for i in range(len(each_tweet)):
        tweet = json.loads(each_tweet[i])
        id = tweet['id']
        retweet = tweet['retweet_count']
        like = tweet['favorite_count']
        rating = tweet['full_text'].split()
        r = [rating[i] for i in range(len(rating)) if '/' in rating[i]]
        r = r[0].split('/')
        rating_num = r[0]
        rating_den = r[1]
               
        tweet_list.append({'id': rating,
                          'retweets':retweet,
                          'favorites': like})
                        
df_twitter = pd.DataFrame(tweet_list, columns = ['id', 'retweets', 'favorites'])        